firefox-ios/Client/Frontend/Home/Homepage Rebuild/HomepageDiffableDataSource.swift

```diff
@@ -80,10 +80,15 @@ final class HomepageDiffableDataSource:
     ///   - textColor: text color from wallpaper configuration
     private func getTopSites(
         with topSitesState: TopSitesSectionState,
-        and textColor: TextColor?
+        and textColor: TextColor?,
+        numberOfCellsPerRow: Int
     ) -> [HomepageDiffableDataSource.HomeItem]? {
         guard topSitesState.shouldShowSection else { return nil }
-        let topSites: [HomeItem] = topSitesState.topSitesData.compactMap { .topSite($0, textColor) }
+        let topSites: [HomeItem] = topSitesState.topSitesData.prefix(
+            topSitesState.numberOfRows * numberOfCellsPerRow
+        ).compactMap {
+            .topSite($0, textColor)
+        }
```

Comment:
🔥 this makes sense to me to have the calculation here + the previous implementation had a bug